src/update_agent/actor.rs

```diff
@@ -491,6 +499,52 @@ impl UpdateAgentInfo {
             .unwrap_or_else(|e| log::error!("failed to register as driver: {}", e))
             .await
     }
+
+    /// Cleanup pending deployment.
+    async fn cleanup_pending_deployment(&self) {
+        let msg = rpm_ostree::CleanupPendingDeployment {};
+        let result = self
+            .rpm_ostree_actor
+            .send(msg)
+            .unwrap_or_else(|e| Err(e.into()))
+            .await;
+        if let Err(e) = result {
+            log::error!("failed to cleanup pending deployment: {}", e)
+        };
+    }
+
+    /// Check that the pending deployment is from the correct update stream.
+    /// If not, clean up the pending deployment.
+    async fn check_stream(&self, state: &mut UpdateAgentState, release: Release) {
+        match self.query_pending_deployment_stream().await {
+            Ok(stream) => {
+                if stream != self.identity.stream {
+                    log::error!(
+                        "deployed an update on different update stream, abandoning update {}",
+                        release.version
+                    );
+                } else {
+                    let msg = format!("update staged: {}", release.version);
+                    utils::update_unit_status(&msg);
+                    log::info!("{}", msg);
+                    state.machine_state.update_staged(release);
+                    // The release we just deployed is on the correct stream so we
+                    // return early.
+                    return;
+                }
+            }
+            Err(e) => {
+                log::error!(
+                    "failed to check pending deployment's update stream: {}, abandoning update {}",
```

Comment:
I think this is a bit too aggressive, on transient errors (e.g. dbus communication failures) we may end up blocking a release through the `denylist` below.